CZERTAINLY Keycloak Theme is not compatible with Keycloak 24

**Describe the bug**

CZERTAINLY Keycloak Theme is not compatible with Keycloak 24. In case user is not know to Keycloak **and** has some incorrect/missing attributes. Keycloak tries to display registration form to get missing values, but that form was customized for Keycloak 21, and is missing something Keycloak 24 requires. Whole login process crashes with  *An internal server error has occurred*

The error won't show in case all user attributes are provided by the IdP and all are valid. In example bellow I show incorect Username with value `3KEY\semik`. The character `\` is no longer valid with Keycloak 24, it was valid  with version 21. This will require further exploration of AD FS integration. But at this moment, we need to resolve that _Internal server error_. 

Explored with @KaterinaBzatkova 

**To Reproduce**
Steps to reproduce the behavior:
1. Configure integration with AD FS, prepare testing user
2. Try to login using AF FS
3. See error

**Expected behavior**

I expect login into CZERTAINLY, not error screen.

**Screenshots**

Example of error screen:
![image](https://github.com/3KeyCompany/CZERTAINLY-Keycloak-Theme/assets/4614807/aeea24c3-c471-4da5-bfc6-f16ae96f9e6c)

Example of official Keycloak registration:
![image](https://github.com/3KeyCompany/CZERTAINLY-Keycloak-Theme/assets/4614807/45cf94b6-a77e-4915-a46b-301d37311e4f)


**Desktop (please complete the following information):**
 - OS: Debian GNU/Linux Bookworm
 - Browser: Firefox  
 - Version: 115.11.0esr (64-bit)

CZERTAINLY Helm Chart 2.12.0, images:
```
api-gateway
  cont: docker.io/revomatico/docker-kong-oidc:3.4.0-2
auth-service
  cont: docker.io/3keycompany/czertainly-auth:1.4.0
auth-service
  cont: docker.io/3keycompany/czertainly-auth:1.4.0
auth-opa-policies
  cont: docker.io/3keycompany/czertainly-auth-opa-policies:1.2.0
common-credential-provider
  cont: docker.io/3keycompany/czertainly-common-credential-provider:1.3.2
common-credential-provider
  cont: docker.io/3keycompany/czertainly-common-credential-provider:1.3.2
czertainly
  init: docker.io/curlimages/curl:8.1.1
  cont: docker.io/openpolicyagent/opa:0.53.0-rootless
  cont: docker.io/3keycompany/czertainly-core:2.12.0
czertainly
  init: docker.io/curlimages/curl:8.1.1
  cont: docker.io/openpolicyagent/opa:0.53.0-rootless
  cont: docker.io/3keycompany/czertainly-core:2.12.0
cryptosense-discovery-provider
  cont: harbor.3key.company/czertainly/czertainly-cryptosense-discovery-provider:1.4.0
cryptosense-discovery-provider
  cont: harbor.3key.company/czertainly/czertainly-cryptosense-discovery-provider:1.4.0
ejbca-ng-connector
  cont: docker.io/3keycompany/czertainly-ejbca-ng-connector:1.9.0
ejbca-ng-connector
  cont: docker.io/3keycompany/czertainly-ejbca-ng-connector:1.9.0
email-notification-provider
  cont: docker.io/3keycompany/czertainly-email-notification-provider:1.0.0
email-notification-provider
  cont: docker.io/3keycompany/czertainly-email-notification-provider:1.0.0
fe-administrator
  cont: docker.io/3keycompany/czertainly-frontend-administrator:2.12.0
keycloak-internal
  init: docker.io/3keycompany/czertainly-keycloak-theme:0.1.1
  cont: docker.io/3keycompany/czertainly-keycloak-optimized:24.0.2-0
keystore-entity-provider
  cont: docker.io/3keycompany/czertainly-keystore-entity-provider:1.4.1
keystore-entity-provider
  cont: docker.io/3keycompany/czertainly-keystore-entity-provider:1.4.1
messaging-service
  cont: docker.io/rabbitmq:3.12.1
ms-adcs-connector
  cont: harbor.3key.company/czertainly/czertainly-ms-adcs-connector:1.6.0
ms-adcs-connector
  cont: harbor.3key.company/czertainly/czertainly-ms-adcs-connector:1.6.0
network-discovery-provider
  cont: docker.io/3keycompany/czertainly-ip-discovery-provider:1.4.0
network-discovery-provider
  cont: docker.io/3keycompany/czertainly-ip-discovery-provider:1.4.0
pyadcs-connector
  cont: harbor.3key.company/czertainly/czertainly-pyadcs-connector:1.1.0
scheduler-service
  init: docker.io/curlimages/curl:8.1.1
  cont: docker.io/3keycompany/czertainly-scheduler:1.0.0
scheduler-service
  init: docker.io/curlimages/curl:8.1.1
  cont: docker.io/3keycompany/czertainly-scheduler:1.0.0
software-cryptography-provider
  cont: docker.io/3keycompany/czertainly-software-cryptography-provider:1.2.2
software-cryptography-provider
  cont: docker.io/3keycompany/czertainly-software-cryptography-provider:1.2.2
utils-service
  cont: harbor.3key.company/czertainly/czertainly-utils-service:1.0.0
utils-service
  cont: harbor.3key.company/czertainly/czertainly-utils-service:1.0.0
x509-compliance-provider
  cont: docker.io/3keycompany/czertainly-x509-compliance-provider:1.3.0
x509-compliance-provider
  cont: docker.io/3keycompany/czertainly-x509-compliance-provider:1.3.0
```

**Additional context**

Error message in Keycloak POD:
```
2024-06-04 10:03:42,196 DEBUG [org.keycloak.saml.validators.ConditionsValidator] (executor-thread-1) Assertion _74c5a69d-96c6-4e6e-92bd-689c25da8b64 validity is VALID
2024-06-04 10:03:42,305 SEVERE [freemarker.runtime] (executor-thread-1) Error executing FreeMarker template: freemarker.core._MiscTemplateException: Macro "registrationLayout" has no parameter with name "displayRequiredFields". Valid parameter names are: bodyClass, displayInfo, displayMessage

----
FTL stack trace ("~" means nesting-related):
	- Failed at: #macro registrationLayout bodyClass="...  [in template "template.ftl" in macro "registrationLayout" at line 1, column 1]
	- Reached through: @layout.registrationLayout displayMes...  [in template "idp-review-user-profile.ftl" at line 3, column 1]
----
	at freemarker.core.Environment.newUndeclaredParamNameException(Environment.java:1143)
	at freemarker.core.Environment.setMacroContextLocalsFromArguments(Environment.java:1023)
	at freemarker.core.Environment.invokeMacroOrFunctionCommonPart(Environment.java:859)
	at freemarker.core.Environment.invokeMacro(Environment.java:813)
	at freemarker.core.UnifiedCall.accept(UnifiedCall.java:84)
	at freemarker.core.Environment.visit(Environment.java:335)
	at freemarker.core.Environment.visit(Environment.java:341)
	at freemarker.core.Environment.process(Environment.java:314)
	at freemarker.template.Template.process(Template.java:383)
	at org.keycloak.theme.freemarker.DefaultFreeMarkerProvider.processTemplate(DefaultFreeMarkerProvider.java:49)
	at org.keycloak.forms.login.freemarker.FreeMarkerLoginFormsProvider.processTemplate(FreeMarkerLoginFormsProvider.java:544)
	at org.keycloak.forms.login.freemarker.FreeMarkerLoginFormsProvider.createResponse(FreeMarkerLoginFormsProvider.java:316)
	at org.keycloak.forms.login.freemarker.FreeMarkerLoginFormsProvider.createUpdateProfilePage(FreeMarkerLoginFormsProvider.java:638)
	at org.keycloak.authentication.authenticators.broker.IdpReviewProfileAuthenticator.authenticateImpl(IdpReviewProfileAuthenticator.java:75)
	at org.keycloak.authentication.authenticators.broker.AbstractIdpAuthenticator.authenticate(AbstractIdpAuthenticator.java:76)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processSingleFlowExecutionModel(DefaultAuthenticationFlow.java:442)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processFlow(DefaultAuthenticationFlow.java:246)
	at org.keycloak.authentication.AuthenticationProcessor.authenticateOnly(AuthenticationProcessor.java:1027)
	at org.keycloak.services.resources.LoginActionsService$1.authenticateOnly(LoginActionsService.java:884)
	at org.keycloak.authentication.AuthenticationProcessor.authenticate(AuthenticationProcessor.java:884)
	at org.keycloak.services.resources.LoginActionsService.processFlow(LoginActionsService.java:366)
	at org.keycloak.services.resources.LoginActionsService.brokerLoginFlow(LoginActionsService.java:914)
	at org.keycloak.services.resources.LoginActionsService.firstBrokerLoginGet(LoginActionsService.java:791)
	at org.keycloak.services.resources.LoginActionsService$quarkusrestinvoker$firstBrokerLoginGet_7740895ae1600a88809add5c32d725c3dfcd8338.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:840)

2024-06-04 10:03:42,308 ERROR [org.keycloak.forms.login.freemarker.FreeMarkerLoginFormsProvider] (executor-thread-1) Failed to process template: org.keycloak.theme.FreeMarkerException: Failed to process template idp-review-user-profile.ftl
	at org.keycloak.theme.freemarker.DefaultFreeMarkerProvider.processTemplate(DefaultFreeMarkerProvider.java:52)
	at org.keycloak.forms.login.freemarker.FreeMarkerLoginFormsProvider.processTemplate(FreeMarkerLoginFormsProvider.java:544)
	at org.keycloak.forms.login.freemarker.FreeMarkerLoginFormsProvider.createResponse(FreeMarkerLoginFormsProvider.java:316)
	at org.keycloak.forms.login.freemarker.FreeMarkerLoginFormsProvider.createUpdateProfilePage(FreeMarkerLoginFormsProvider.java:638)
	at org.keycloak.authentication.authenticators.broker.IdpReviewProfileAuthenticator.authenticateImpl(IdpReviewProfileAuthenticator.java:75)
	at org.keycloak.authentication.authenticators.broker.AbstractIdpAuthenticator.authenticate(AbstractIdpAuthenticator.java:76)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processSingleFlowExecutionModel(DefaultAuthenticationFlow.java:442)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processFlow(DefaultAuthenticationFlow.java:246)
	at org.keycloak.authentication.AuthenticationProcessor.authenticateOnly(AuthenticationProcessor.java:1027)
	at org.keycloak.services.resources.LoginActionsService$1.authenticateOnly(LoginActionsService.java:884)
	at org.keycloak.authentication.AuthenticationProcessor.authenticate(AuthenticationProcessor.java:884)
	at org.keycloak.services.resources.LoginActionsService.processFlow(LoginActionsService.java:366)
	at org.keycloak.services.resources.LoginActionsService.brokerLoginFlow(LoginActionsService.java:914)
	at org.keycloak.services.resources.LoginActionsService.firstBrokerLoginGet(LoginActionsService.java:791)
	at org.keycloak.services.resources.LoginActionsService$quarkusrestinvoker$firstBrokerLoginGet_7740895ae1600a88809add5c32d725c3dfcd8338.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: freemarker.core._MiscTemplateException: Macro "registrationLayout" has no parameter with name "displayRequiredFields". Valid parameter names are: bodyClass, displayInfo, displayMessage

----
FTL stack trace ("~" means nesting-related):
	- Failed at: #macro registrationLayout bodyClass="...  [in template "template.ftl" in macro "registrationLayout" at line 1, column 1]
	- Reached through: @layout.registrationLayout displayMes...  [in template "idp-review-user-profile.ftl" at line 3, column 1]
----
	at freemarker.core.Environment.newUndeclaredParamNameException(Environment.java:1143)
	at freemarker.core.Environment.setMacroContextLocalsFromArguments(Environment.java:1023)
	at freemarker.core.Environment.invokeMacroOrFunctionCommonPart(Environment.java:859)
	at freemarker.core.Environment.invokeMacro(Environment.java:813)
	at freemarker.core.UnifiedCall.accept(UnifiedCall.java:84)
	at freemarker.core.Environment.visit(Environment.java:335)
	at freemarker.core.Environment.visit(Environment.java:341)
	at freemarker.core.Environment.process(Environment.java:314)
	at freemarker.template.Template.process(Template.java:383)
	at org.keycloak.theme.freemarker.DefaultFreeMarkerProvider.processTemplate(DefaultFreeMarkerProvider.java:49)
	... 24 more

2024-06-04 10:03:42,308 ERROR [org.keycloak.headers.DefaultSecurityHeadersProvider] (executor-thread-1) MediaType not set on path /kc/realms/CZERTAINLY/login-actions/first-broker-login, with response status 500
2024-06-04 10:03:42,309 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-1) Uncaught server error: jakarta.ws.rs.InternalServerErrorException: HTTP 500 Internal Server Error
	at org.keycloak.headers.DefaultSecurityHeadersProvider.addHeaders(DefaultSecurityHeadersProvider.java:75)
	at org.keycloak.services.filters.KeycloakSecurityHeadersFilter.filter(KeycloakSecurityHeadersFilter.java:43)
	at org.jboss.resteasy.reactive.server.handlers.ResourceResponseFilterHandler.handle(ResourceResponseFilterHandler.java:25)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:150)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

**Workaround to get login working**
Delete `InitContainer` from `keycloak-internal-deployment`.

--- a/custom_themes/themes/czertainly/login/template.ftl
+++ b/custom_themes/themes/czertainly/login/template.ftl
@@ -1,4 +1,4 @@
-<#macro registrationLayout bodyClass="" displayInfo=false displayMessage=true>
+<#macro registrationLayout bodyClass="" displayInfo=false displayMessage=true displayRequiredFields=false>
 <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
 <html xmlns="http://www.w3.org/1999/xhtml">
 
@@ -29,7 +29,7 @@
         </div>
         </#if>
         <#nested "form">
-            </div> 
+            </div>
         </div>
 	</body>
 </html>

